### PR TITLE
Add query and candidate

### PR DESCRIPTION
**Pull request by sahel-sh**
This cl:
- adds rerank_batch, query and candidate data classes.
- removes doc inline demo
- retriever only supports dataset or custom index,
- pyserini_retriever creates requests
- updates all demos to work with rerank_batch
- rerank (non-batch) will be added in a follow up cl
- tests will be updated in a follow up cl
- if we decide to keep pyserini retriever, caching will be reenabled in a followup cl.

# Pull Request Checklist

## Reference Issue

Please provide the reference to issue this PR is addressing (# followed by the issue number). If there is no associated issue, write "N/A".

ref: #110 

## Checklist Items

Before submitting your pull request, please review these items:

- [ ] Have you followed the [contributing guidelines](CONTRIBUTING.md)?
- [x] Have you verified that there are no existing [Pull Requests](https://github.com/castorini/rank_llm/pulls) for the same update/change?
- [ ] Have you updated any relevant documentation or added new tests where needed?

# PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... 
    - Description:

### CODE REVIEW

File: src/rank_llm/data.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                candidates = [candidate.__dict__ for candidate in d.candidates]
+                json.dump({"query": d.query.__dict__, "candidates": candidates}, f)
+                f.write("\n")
+
+    def write_in_trec_eval_format(self, filename: str):
+        with open(filename, "a" if self._append else "w") as f:
+            for d in self._data:
+                qid = d.query.qid
+                for rank, cand in enumerate(d.candidates, start=1):
+                    f.write(f"{qid} Q0 {cand.docid} {rank} {cand.score} rank\n")
```
Comment (ronakice) [inline src/rank_llm/data.py:101]: maybe `rank_llm` here as the tag. Ideally we rework this in the future to be specific to model.

File: src/rank_llm/demo/rerank_inline_hits.py [hunk 2]
```diff
@@ -1,85 +1,86 @@
 import os
+from pathlib import Path
 import sys
 
+from dacite import from_dict
```
Comment (ronakice) [inline src/rank_llm/demo/rerank_inline_hits.py:5]: Is this dependency added somewhere?

File: src/rank_llm/evaluation/trec_eval.py [hunk 3]
```diff
@@ -35,10 +35,9 @@ def from_results(
         ).name
         with open(temp_run_file, "w") as file:
             for result in results:
-                for hit in result.hits:
-                    file.write(
-                        f"{hit['qid']} Q0 {hit['docid']} {hit['rank']} {hit['score']} rank\n"
-                    )
+                qid = result.query.qid
+                for rank, cand in enumerate(result.candidates, start=1):
+                    file.write(f"{qid} Q0 {cand.docid} {rank} {cand.score} rank\n")
```
Comment (ronakice) [inline src/rank_llm/evaluation/trec_eval.py:40]: `rank_llm` here too?

**Review by ronakice — CHANGES_REQUESTED** [review]
Some minor nits but looks good to me besides

File: src/rank_llm/demo/rerank_inline_hits.py (hunk 2 above)
Comment (sahel-sh) [reply src/rank_llm/demo/rerank_inline_hits.py:5]: Added to requiremens.txt

File: src/rank_llm/data.py (hunk 1 above)
Comment (sahel-sh) [reply src/rank_llm/data.py]: done

File: src/rank_llm/evaluation/trec_eval.py (hunk 3 above)
Comment (sahel-sh) [reply src/rank_llm/evaluation/trec_eval.py]: done

**Review by ronakice — APPROVED** [review]
LGTM!

**Pull request merged**